src/main/java/frc/robot/subsystems/shooter/Shooter.java

```diff
@@ -66,21 +126,59 @@ else if (stateSpaceType.equals("Inertia"))
                 NOMINAL_VOLTAGE, LOOP_PERIOD);
     }
 
+    /**
+     * Gets the velocity of the motor.
+     *
+     * @return the velocity of the motor. [rpm]
+     */
     public double getVelocity() {
-        return unitModel.toUnits(mainMotor.getSelectedSensorVelocity());
+        if (Robot.isSimulation()) {
+            return flywheelSim.getAngularVelocityRPM();
+        }
+        return unitModel.toVelocity(motor.getSelectedSensorVelocity());
     }
 
-    public void setVelocity(double velocity, double timeInterval) {
-        velocity = Utils.deadband(velocity, WPI_TalonFX.kDefaultSafetyExpiration);
-
-        linearSystemLoop.setNextR(VecBuilder.fill(velocity));
-        linearSystemLoop.correct(VecBuilder.fill(getVelocity()));
-        linearSystemLoop.predict(timeInterval);
+    /**
+     * Sets the velocity of the motor.
+     *
+     * @param velocity is the velocity setpoint. [rpm]
+     */
+    public void setVelocity(double velocity) {
+        linearSystemLoop.setNextR(VecBuilder.fill(Units.rotationsToRadians(Utils.secondsToMinutes(velocity))));
+        linearSystemLoop.correct(VecBuilder.fill(Units.rotationsToRadians(Utils.secondsToMinutes(getVelocity()))));
+        linearSystemLoop.predict(currentTime - lastTime);
+        if (Robot.isSimulation()) {
+            SmartDashboard.putNumber("ve", velocity);
+            flywheelSim.setInputVoltage(MathUtil.clamp(linearSystemLoop.getU(0), -NOMINAL_VOLTAGE, NOMINAL_VOLTAGE));
+            simCollection.setIntegratedSensorVelocity(unitModel.toTicks100ms(Utils.secondsToMinutes(flywheelSim.getAngularVelocityRPM())));
+        } else {
+            motor.setVoltage(MathUtil.clamp(linearSystemLoop.getU(0), -NOMINAL_VOLTAGE, NOMINAL_VOLTAGE));
+        }
+    }
 
-        mainMotor.set(ControlMode.PercentOutput, linearSystemLoop.getU(0) / NOMINAL_VOLTAGE);
+    public void setPower(double output) {
+        motor.set(ControlMode.PercentOutput, output);
     }
 
+    /**
+     * Terminates the movement of the wheel.
+     */
     public void terminate() {
-        mainMotor.set(ControlMode.PercentOutput, 0);
+        motor.stopMotor();
+    }
+
+    @Override
+    public void periodic() {
+        lastTime = currentTime;
+        currentTime = Timer.getFPGATimestamp();
+    }
+
+    @Override
+    public void simulationPeriodic() {
+        lastTime = currentTime;
+        SmartDashboard.putNumber("velocity", flywheelSim.getAngularVelocityRPM());
+        SmartDashboard.putNumber("power", simCollection.getMotorOutputLeadVoltage());
+        flywheelSim.update(LOOP_PERIOD);
+        currentTime = Timer.getFPGATimestamp();
```

Comment:
please group currentTime with lastTime

Reply:
Wdym?

Reply:
Just move`currentTime = ...` to be right after `lastTime`